### ISSUE TITLE
Feature Request: C++ gradient for Floor

### ISSUE THREAD

**Issue by gdeer81** [tensorflow/tensorflow#13982]
Implement the gradient for Floor in c++ so that it is available for TF_AddGradients

@suharshs I will implement this one with some guidance from @bpiel 

I believe this is the python code I'll be porting over:
https://github.com/tensorflow/tensorflow/blob/962ed613cf1087637848d3e2b23f5b01d93c7eda/tensorflow/python/ops/math_grad.py#L1004-L1006

**Comment by bpiel** [comment]
cc @skyewm

**Comment by bpiel** [comment]
cc @skye

**Comment by gdeer81** [comment]
the python implementation uses None to indicate "no connection" but I'm not sure that returning null in c++ will have the same effect.  

So I have three options: return 0, return the input gradient, or return a zero matrix. 

Also, there is no test for the python version, so I'm really lost on how to test this one.

**Comment by skye** [comment]
Use NoGradient() (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/cc/framework/gradients.h#L48). For example: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/cc/gradients/nn_grad.cc#L193

As for testing, look at cc/gradients/math_grad_test.cc and see if you can add a test case to CWiseUnaryGradTest for Floor. If you still need help post back here (I haven't looked at these tests myself so I can't offer guidance off the top of my head, but it looks like there are plenty of examples).

**Comment by theflofly** [comment]
@gdeer81 Are you working on that ?

**Comment by bpiel** [comment]
@theflofly I know he was, but I haven't heard from him (on slack) since he posted that last comment (Nov 6).

**Comment by theflofly** [comment]
Thanks @bpiel, is this a slack related to TF developments ?

**Comment by bpiel** [comment]
@theflofly The slack I was referring to above is the tensorflow channel of the clojure slack (https://clojurians.slack.com/), but since you asked, I've also made a channel in a TF-focused slack for non-python TF collaboration. It doesn't see a lot of action, but if you're interested, please email me. I don't want to hijack this issue.

**Comment by gdeer81** [comment]
@theflofly I haven't made much progress with this one if you're looking for a feature request that's probably in your wheelhouse.

**Comment by theflofly** [comment]
@gdeer81 you should it, it is easy enough to get you started. What is your problem ?

Something like this should work:
```
Status FloorGrad(const Scope& scope, const Operation& op,
               const std::vector<Output>& grad_inputs,
               std::vector<Output>* grad_outputs) {
  grad_outputs->push_back(NoGradient());
  return scope.status();
}
REGISTER_GRADIENT_OP("Floor", FloorGrad);
```

**Comment by gdeer81** [comment]
@theflofly that's exactly what I have. =D
I was struggling to write the test. 
Someone said to look at your PR for ProdGrad and it was kind of intimidating so I was worried that I left something out.

I'll push what I have tonight let everyone see what I've done. Sunlight is the best disinfectant.

**Comment by gdeer81** [comment]
This is what I have so far:
https://github.com/gdeer81/tensorflow/commit/5275fc4dfd584ea3e95ea5227259c4e9d36d2f13

It took a bit of fiddling to get the test to go from erroring out to just plain failing. Using the x_fn variable like the other tests do just caused a seg fault so I commented it out and just started plugging in random values for the second arg to the test method. The only acceptable value so far has been 0.

But the test output doesn't say why the test failed so I'm kind of at a loss for how to write a proper test for this one despite all the examples  whose test inputs look arbitrary and don't really provide any helpful information to anyone who is trying to write new tests.

**Comment by theflofly** [comment]
@gdeer81 I'll check and come back to you.

**Comment by gdeer81** [comment]
@theflofly any thoughts on how to test my changes?

**Comment by theflofly** [comment]
@gdeer81 Sorry for the delay. 

As a NoGradient cut the flow of the gradient I don't see what we can test. I mean if you do something like that:
```
auto y = Square(Floor(Matmul(x1, X2)))
```
And you try to differentiate y w.r.t x1 or x2, it will just do nothing (no error, no result) whereas in Python there is:  `No gradients provided for any variable, check your graph for ops that do not support gradients, between variables...`. I think that the warning should be also added, else it will be very difficult to debug.

Also I think that:
```
REGISTER_NO_GRADIENT_OP("Floor");
```
is enough, no need for the code that I previously pasted, NoGradientOp are already defined like that for LogicalNot, LogicalOr, ... and they aren't tested either.

In summary, I'd say add only `REGISTER_NO_GRADIENT_OP("Floor");` and we should test if the variable are reachable taking into account NoGradient links. As I worked on reachable nodes I can eventually do it, what do you think @skye ?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by theflofly** [comment]
@skye Can you check my last comment please ?

**Comment by skye** [comment]
Sorry for the delay, I missed this and then went on vacation.

@theflofly I think you're right and we can just use REGISTER_NO_GRADIENT_OP, sorry for not catching that earlier! And no need to test in that case, we should already have test coverage for gradientless ops.

**Comment by theflofly** [comment]
@skye No problem. Shouldn't we add a check if the graph is not cut by a NoGradient op in order to display a helpful message ?

**Comment by skye** [comment]
Ah, I think I didn't fully parse your previous comment. Are you suggesting we should add a similar "No gradients provided for any variable" error to the C++ gradient code? If so I disagree, this error is actually generated by Optimizer, not the low-level gradients code.

**Comment by theflofly** [comment]
@skye Sorry for the delay. I agree with you, we'll add that when the optimizers will be available in C++.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @gdeer81: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @gdeer81: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @gdeer81: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by theflofly** [comment]
I think so.

**Comment by skye** [comment]
I think so as well.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @gdeer81: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @gdeer81: It has been 31 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by theflofly** [comment]
I created a pull request https://github.com/tensorflow/tensorflow/pull/19662 so that we can close this issue.
